### ISSUE TITLE
Message Bus for Stratego4

### ISSUE THREAD

**Issue by barlock** [ConsenSys/web3studio-sidejam#1]
## Overview

This is an Epic encapsulating research on using an EVM as a message bus and then applying that concept to the game of Stratego4. 

Stratego4 was chosen as a simple example of cross-company blockchain patterns:
1. Information that all parties know and share
2. Information that each party keeps secret to themselves
3. Information shared between a subset of the parties
4. Shared information that can eventually be proven true or false 

### Reference

- [Atomic Crosschain Transactions (Pantheon)](https://arxiv.org/abs/1904.12079)
- [chain.link](https://docs.chain.link/docs)

### Assumptions

- Even Stratego4's rules can be complex, the rules can and likely should be modified to produce a simpler, better, metaphor 

### Acceptance

- Documented architecture pattern for a Message Bus on an EVM
  - Should help to prevent deadlocks
  - Should help prevent double spend
  - Should attempt to document failure states
- A Playable Stratego4 using a message bus

**Comment by barlock** [comment]
Storyboard of Gameplay: https://app.mural.co/t/consensys3989/m/consensys3989/1557345438344/718181f0ee878f1cfdc2756c9dc6aa363680b634

![image](https://user-images.githubusercontent.com/5770007/57456734-16584a80-723c-11e9-904e-565cd9eddc17.png)

**Comment by barlock** [comment]
Simple Message Bus pattern
![image](https://user-images.githubusercontent.com/5770007/57457412-7e5b6080-723d-11e9-8bac-dc5fecfda334.png)

**Comment by barlock** [comment]
![image](https://user-images.githubusercontent.com/5770007/57459274-e8c1d000-7240-11e9-8c10-8e4dd73210c6.png)

**Comment by barlock** [comment]
Official rules for Stratego4

## Introduction 
You are a military commanding officer on the eve of a mighty battle. The fortress you must capture looms in the distance. But there are two, or even three, opponents and they, too, will fight to call the fortress their own! 

## The battle takes place with: 
one double-sided board 
4 sets of figures in 4 colours 
4 sets of cardboard ranks 
4 flags and stickers in 4 colours 
4 promotion cards 
I fortress 

# Before playing for the first time: 
Attach the ranks to the figures, taking care to use the correct colour (see illustration). Then place the stickers on the flags. 

## Before the battle starts: 
The board has a side for three players and a side for four players. Place the appropriate side up and put the fortress in the middle. 
Choose a colour and take your place at the appropriate side of the board. Then take the corresponding army, promotion card and flag. Set the moveable part of the promotion card to '0' 

## The object of the game: 
You must get your flag and one of your pieces to the top of the fortress. The first player to succeed is the winner and the battle is over. 

## How to set up your army: 
Place each of your 20 pieces on fields of the same colour. The centre field (see illustration on page 8) should be left empty. The pieces should be positioned in such a way that only you, and not your opponents, 
can see their ranks. 
Attach your flag to any piece. 
This piece must staff on the field illustrated with flags. 
You will learn from experience how best to arrange your pieces. Start by spreading out your high and low pieces and do not put the Cannons right at the back. You may need them in a hurry! 

## How to play: 
The players decide who is to start and then play in turn. You can either move a piece or attack one of your opponent s pieces. 
Only one move is allowed per turn. 

### Moving the pieces: 
1. Pieces may be moved one field to the right, to the left, forwards or backwards (there are some exceptions, which will be discussed later). 
2. Pieces may not be moved diagonally; only one piece may occupy a field at any time and pieces may not jump over other pieces. 
3. Pieces may not move into the lakes, nor can they jump over them. Cannons can, of course, be fired across them! 
4. The fortress in the middle has two fields. the first ring and the top. The first ring counts as a single field, so only one piece may occupy it at any given moment. You can move into it from any of the 12 surrounding fields, and move out of it into any of these surrounding fields. A piece in the first ring can also be attacked by a piece from any of these 12 fields - and the other way around! 
5. A piece may not move to and fro between two adjacent fields more than five times. 
It is important to remember which player started moving to and fro, because this 
player has to stop doing it first and this can mean losing an important piece. 
6. The central field in the area m front of you is called the promotion field. 
Its purpose will be described more fully later on. Leave it empty when you are 
setting out your pieces, although all pieces are allowed to use this field during play. 

### Attacking: 
1. If an opponent's piece is directly in front of you, next to you or behind you, you can attack it; but you don't have to if you don't want to. 
2. To attack an opponent: take your piece, tap your opponent's piece with it, and state your rank. Your opponent then names his rank. 
3. The piece with the lowest rank loses and is removed from the board. 
If the attacking piece wins, it takes the place of the losing one; if the defending 
piece wins, it stays put. 
4. If the attacking piece and the defending piece turn out to have the same rank, both lose and both are removed. 
5. Losing pieces return to their owner, who must arrange them in front of him in order of rank so that all the other players can see these ranks. 

### Your pieces: 
The ranks are displayed in correct order on your promotion card. To make them easier 
to tell apart, each rank also has a number. The Marshal is the highest-ranking piece 
and has the number 9; the General is a number 8, and so on, down to the Cannon 
with the number 1. A Marshal beats a General and all other lower ranks, a 
General beats a Major and all other lower ranks, and so on, down to the Cannon, 
which can be beaten by any other rank. 

### A few ranks have special features: 
**Captains** belong to the cavalry, so they can move faster. A Captain is allowed to move two fields at once, changing direction half- 
way if he wants to. 
He can also attack opponents only once during such a move, though, and if he attacks an opponent in the first field, then he cannot move any further in that turn. 

**Scouts** can jump over more than one empty field in a straight line; left, right, forwards and backwards. 
They cannot jump over pieces, over the fortress, or over the lakes. So a Scout can 
attack from a great distance, provided there are only empty fields in between. 
Scouts can mean real trouble for Cannons and Spies! 

**Spies** are the same deceptive characters they were in the original Stratego. 
Almost every piece that attacks a Spy, wins. But If a Spy attacks a Marshal, the 
Marshal dies! The Spy has to attack first in order to kill a Marshal; if the Marshal 
attacks the Spy first, then the Spy dies. 

**Cannons** are dangerous weapons. 
Think carefully about when you fire a Cannon and where you aim it, because a 
Cannon can only be fired once; after it is fired it is removed from the board. 
Look after them, because as the lowest ranking piece Cannons can be beaten by 
every other piece - including from a distance, by a Scout! 
They are moved just like ordinary pieces. 
A fired Cannon beats every other piece, but their field of fire is limited: three fields, in a straight line. So the cannonball lands not one, or two, but exactly three fields away. 
And the two fields in between have to be empty! 
Cannons can fire over lakes; a lake then counts as one field. Cannons cannot fire over the fortress. 

### The flag: 
Any piece can carry the flag. Just fix the flag onto it. But any piece holding the flag, even a Scout or a Captain, can only move one field at a time. 
If the piece holding your flag is beaten, your opponent can take your flag, and move 
away with it in a later turn. If this happens, you will have to win your flag back to be able to win the game! 
If your flag is attacked by an opponent's piece of the same rank, or if it is attacked by a Cannon, then both pieces are removed from the board but the flag remains, unguarded, on the ground. 
The first piece to arrive at that field can pick up the flag. 
If you want, you can drop the flag and let another one of your pieces pick it up on 
your next turn. A piece can only carry a single flag. 

### Promotion: 
At the beginning of the game, each player gets a promotion card and sets the slider to Every time you attack an enemy piece or are attacked, and you beat that piece, 
you can slide one number up. When you reach the 'VI', you must choose one of your 
pieces that was beaten earlier and put it back on the board, in the promotion field. 
You have to choose straight away, and if your promotion field is alrea4y occupied, 
then tough luck - you lost your chance! 

![image](https://user-images.githubusercontent.com/5770007/57460492-37706980-7243-11e9-8b0b-0a58afc09992.png)

Google Translated:
That's where they are the strongest.
When you place a cannon behind one of ponds, the first circle of the castle is
then find the fire range!
In other words, it's the best place for post your cannons.
When you play a 3 and your two opponents attack you, it will be hard for you
to defend you for a long time
If your cannon is li, you can not shoot in the direction strong castle. Indeed, you do not have not the right to shoot over castle.

Then try to convince one to lead its flag in the direction of the strong castle or
begin to do it yourself. Be careful not to waste your highs officers; however, there is no point in being as careful as when you play a
Stratego Original.

Thanks to the promotion, you can indeed recover one of your pieces!
3. A scout that can be found can access these boxes.
The checkbox
This box is where
🎌The piece that carries the flag.
